### ISSUE TITLE
Issue with Chef Client 13  and the `chef_client_scheduled_task` resource on Windows

### ISSUE THREAD

**Issue by seththoenen** [chef-cookbooks/chef-client#490]
### Cookbook version
chef-client >= 8.1.0

### Chef-client version
chef-client 13.0.118

### Platform Details
Windows Server 2008 R2
Windows Server 2012 R2

### Scenario:
The `chef_client_scheduled_task` resource converges on every run on chef-client cookbook version >= 8.1.0 when paired with chef-client 13.0.118 on Windows Server 2008 R2 and Windows Server 2012 R2. Currently utilizing the windows cookbook version 3.0.5 (the latest at the time of this writing). Please refer to the table below:

![image](https://cloud.githubusercontent.com/assets/524558/26165388/67ab24c8-3af5-11e7-92d6-4dcde0af6c77.png)

### Steps to Reproduce:
Run the chef-client cookbook version >= 8.1.0 on Windows Server 2008 R2 or Windows Server 2012 R2 paired with the windows cookbook = 3.0.5 and chef-client 13.0.118

### Expected Result:
The `chef_client_scheduled_task` resource should not converge on every chef-client run.

### Actual Result:
The `chef_client_scheduled_task` resource converges on every chef-client run.

**Comment by iennae** [comment]
8.0.2 had known issues with working on Windows systems. We'll dig into the issue with 8.1.0 and Chef 13. Thanks for reporting this issue.

**Comment by iennae** [comment]
This could be due to a idempotence issue with the windows_task resource https://github.com/chef/chef/issues/6093. 

Could you provide more information about how you are specifying the resource itself? Are you specifying the start time?

**Comment by seththoenen** [comment]
I'm using the default attributes that come with the chef-client cookbook. I'm not overriding any functionality or calling any resources directly at this time. I'm simply calling the cookbook:

`include_recipe 'chef-client::default'`

So, to answer your question, I'm not specifying  a start time.

Thanks!

**Comment by username-is-already-taken2** [comment]
I noticed this frozen error today with 2012 R2 on an Azure VM installed with the chef extension.  I'll try upgrading the cookbook

**Comment by seththoenen** [comment]
I'm still having issues with the latest version of the Chef-client cookbook (version 8.1.6). The Windows scheduled task will still converge. I'm using the default settings that come with the cookbook.

Chef-Client version: 13.1.31

Windows Server Versions:
2008 R2 Standard
2008 R2 Enterprise
2012 Standard
2012 R2 Standard
2016 Standard
2016 Datacenter

**Comment by tas50** [comment]
@seththoenen Can you try the release candidate of Chef 13.3. We did a bunch of work in 13.2 and 13.3 to improve the idempotency of the scheduled task resource.

**Comment by seththoenen** [comment]
@tas50 thanks for the update! I'm using Chef-client 13.3.42 and the chef-client cookbook version 8.1.8. The chef_client_scheduled_task resource is still converging (Windows Server 2012 R2). 

I'm using the stock settings for the chef-client cookbook, aside from setting a custom interval and splay.

**Comment by tas50** [comment]
Sounds like it's time for me to dig in. Thanks for checking

**Comment by robbAtEBSCO** [comment]
I've also hit this issue on Chef-Client 13.3.42.  Have you been able to reproduce?

**Comment by dheerajd-msys** [comment]
This issue could potentially be resolved by the fix for `windows_task` resource in https://github.com/chef/chef/pull/6312 since `chef-client` cookbook uses `start_time` by default here https://github.com/chef-cookbooks/chef-client/blob/master/recipes/task.rb#L42.

**Comment by dheerajd-msys** [comment]
Please ignore my above comment. I made a fix here https://github.com/chef-cookbooks/chef-client/pull/522

**Comment by seththoenen** [comment]
I just tested this on Chef Client 13.5, when specifying the frequency as `minute` the resource will still update each time.

**Comment by tas50** [comment]
We're in a bit of bind with the cookbook right now that we need to figure out within chef-client itself:

Chef-client the in 13.5 fails when you don't specify a start time for your scheduled task.
Chef-client the cookbook sets a start time of now, but this results in the cookbook changing the start time of the existing scheduled task on every run. In previous chef-client (the app) releases we actually ignored changing the start time. We're doing the right thing now by honoring changes in start time and updating those, but it's causing issues.

We're going to make sure we get chef-client (the app) doing the right thing here and get a release out that works. Until then I'm you're unfortunately stuck with a resource that converges on every run.

**Comment by seththoenen** [comment]
@tas50, thanks for the update! Right now we've deployed `chef-client` 13 to our entire fleet of Windows systems in the past two months. With the `chef-client` cookbook running we are having resources converge every run. Eventually, we'll be rolling out the `chef_client_updater` cookbook to keep the `chef-client` version up to date. However we'll be waiting on a `chef-client` and `chef-client` cookbook version that make the run idempotent before proceeding further.

If you need someone to test any thing on this front, just let me know!

**Comment by tas50** [comment]
@seththoenen I'm going to dig into the windows_task resource today and come up with a stabilization action plan so we can make sure we get it to the point where it behaves as expected with this cookbook. Thanks for opening this and the enormous detail in the table. That certainly helped.

**Comment by seththoenen** [comment]
@tas50 sounds great! Let me know if there's any way I can help or test!

**Comment by username-is-already-taken2** [comment]
@seththoenen - if you don't mind me asking with upgrading your windows host did you opt for upgrading the client as part of the chefrun? or did you go down the route of creating a scheduled task to do it?

I've found upgrading via the chef-run a little hit or miss (this could also be dependent of the version of chef-client we are upgrading from as they are mixed)

I haven't tried the schedule task route yet but was going to try it on Friday.  Any advice would be greatly appreciated :)

**Comment by seththoenen** [comment]
@username-is-already-taken2, we haven't solidified our strategy for upgrading `chef-client` at this time. I've used the `chef_client_updater` cookbook several months ago in my lap and my mileage varied (it was an older version before a bunch of the windows fixes). I've also played around with creating my own recipe to do it. Basically what I did was I wrote a resource that laid down a PowerShell script on the local machine that would download and install `chef-client`. I'd then start a background process in PowerShell and tell it to wait for the current `chef-client` run to end. At that time it would do it's thing outside of a Chef run. It worked alright, but wasn't 100%. There were a few times where i had a system that didn't come back. This was all in my lab, I haven't tried any of this at scale yet.
